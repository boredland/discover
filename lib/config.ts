type Architecture = 'x86_64' | 'aarch64';

export type RepoConfig = {
    url: string;
    name: string;
    repo?: string[];
    branch?: string[];
    arch?: Architecture[];
}

const explode = (config: RepoConfig, instruction: { key: string, value?: string[]}) => {
    if (!instruction.value) return [config]

    return instruction.value.map(v => {
        return { ...config, [instruction.key]: [v] }
    })
}

export const getDBUrl = (repoConfig: RepoConfig) => {
    const levels = [
        { key: 'repo', value: repoConfig.repo }, 
        { key: "branch", value: repoConfig.branch }, 
        { key: "arch", value: repoConfig.arch }]
        .filter(({ value }) => value?.length ?? 0 > 0)
        .sort((a,b) => {
            return (a.value?.length ?? 0) > (b.value?.length ?? 0) ? 1 : -1;
        });

    let configs: RepoConfig[] = [repoConfig];

    for (const level of levels) {
        configs = configs.map(config => { 
            return explode(config, level)
        }).flat()
    }

    const urls: string[] = [];
    for (const config of configs) {
        let url = config.url;

        url = url.replace(/\$name/g, config.name)

        if (config.repo && config.repo[0])
            url = url.replace(/\$repo/g, config.repo[0])

        if (config.branch && config.branch[0])
            url = url.replace(/\$branch/g, config.branch[0])

        if (config.arch && config.arch[0]) {
            url = url.replace(/\$arch/g, config.arch[0])
        }

        urls.push(url)
    }
    return urls
}
