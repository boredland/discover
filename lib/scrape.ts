import https from 'node:https';
import tar from 'tar-stream';
import zlib from 'zlib';

export type Desc = {
  FILENAME: string;
  URL?: string;
  VERSION: string;
  LICENSE?: string;
  ARCH: string;
  BUILDDATE?: string;
  PACKAGER?: string;
  DEPENDS?: string[];
  MAKEDEPENDS?: string[];
  DESC?: string;
  MD5SUM?: string;
  SHA256SUM?: string;
  PGPSIG?: string;
  CSIZE: string;
  ISIZE: string;
  NAME: string;
}

export const parseDesc = (descString?: string) => {
  if (!descString) return;
  try {
    const content = descString
      .split(/\n\n/)
      .filter((el) => el.trim() !== "").filter(el => el.startsWith('%'));

    const map = content.map((el) => {
      const split = el.split('\n')
      const heading = split[0].substring(
        split[0].indexOf("%") + 1, 
        split[0].lastIndexOf("%")
      );
      const values = split.slice(1);
      return [heading, values.length === 1 ? values[0] : values]
    })

    const result = Object.fromEntries(map) as Desc;

    return result;
  } catch (e) {
    throw new Error(`error parsing desc\n` + descString)
  }
}


const getInformation = async (url: string) => {
  const extract = tar.extract()
  const chunks: Desc[] = []

  extract.on('entry', function(header, stream, next) {
    const parts: Buffer[] = []

    if (header.name.endsWith('/desc')) {
      stream.on('data', function(chunk: Buffer) {
          if (chunk)
            parts.push(chunk)
      });
    }  
    stream.on('end', function() {
      const content = parseDesc(Buffer.concat(parts).toString())
      if (content)
        chunks.push(content)
      next()
    })
    stream.resume()
  })
  
  return new Promise<Desc[] | void>(function (resolve) {
    https.get(url,{ headers: { 'accept-encoding': 'gzip' }}, function (response) {
      response
        .pipe(zlib.createGunzip())
        .pipe(extract)
        .on("finish", () => resolve(chunks))
        .on("close", () => resolve(chunks))
    });
  });
}

export const scrape = async (downloadUrl: string) => {
    const info = await getInformation(downloadUrl);
    return info;
}
