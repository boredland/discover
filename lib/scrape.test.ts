import { RepoConfig } from "./config";
import { parseDesc, scrape } from "./scrape";
import fsPromises from 'node:fs/promises'

test("manjaro sway aarch64", async () => {
    const config: RepoConfig = { url: "https://packages.manjaro-sway.download/", name: "manjaro-sway", architecture: 'aarch64' }
    const result = await scrape(config);
    expect(result).not.toBeUndefined();
    expect(result?.length).toBeGreaterThan(10);
})

test("manjaro sway x86", async () => {
    const config: RepoConfig = { url: "https://packages.manjaro-sway.download/", name: "manjaro-sway", architecture: 'x86_64' }
    const result = await scrape(config);
    expect(result).not.toBeUndefined();
    expect(result?.length).toBeGreaterThan(10);
})

test("manjaro community x86", async () => {
    const config: RepoConfig = { url: "https://ftp.gwdg.de/pub/linux/manjaro/", name: "manjaro", architecture: 'x86_64', branch: "stable", repo: "community" }
    const result = await scrape(config);
    expect(result).not.toBeUndefined();
    expect(result?.length).toBeGreaterThan(8000);
})

test("parse desc", async () => {
    const descString = await fsPromises.readFile(__dirname + '/__fixtures__/calamares.desc', 'utf-8');
    const result = parseDesc(descString)
    expect(result).toMatchSnapshot()
})
