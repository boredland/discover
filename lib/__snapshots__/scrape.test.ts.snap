// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse desc 1`] = `
Object {
  "ARCH": "x86_64",
  "BASE": "calamares",
  "BUILDDATE": "1638613406",
  "CSIZE": "4930404",
  "DEPENDS": Array [
    "kconfig",
    "kcoreaddons",
    "kiconthemes",
    "ki18n",
    "kio",
    "solid",
    "yaml-cpp",
    "kpmcore>=4.2.0",
    "mkinitcpio-openswap",
    "boost-libs",
    "ckbcomp",
    "hwinfo",
    "qt5-svg",
    "polkit-qt5",
    "gtk-update-icon-cache",
    "plasma-framework",
    "qt5-xmlpatterns",
    "squashfs-tools",
    "libpwquality",
    "appstream-qt",
    "icu",
  ],
  "DESC": "Distribution-independent installer framework",
  "FILENAME": "calamares-3.2.48-1-x86_64.pkg.tar.zst",
  "ISIZE": "10185320",
  "LICENSE": "LGPL",
  "MAKEDEPENDS": Array [
    "extra-cmake-modules",
    "qt5-tools",
    "qt5-translations",
    "git",
    "boost",
    "",
  ],
  "MD5SUM": "9b34e11331399af3cd84db14006c470e",
  "NAME": "calamares",
  "PACKAGER": "Unknown Packager",
  "PGPSIG": "iQGzBAABCgAdFiEEpExkTXknZ87XlBr+q7IHXV8xDPgFAmGrRKQACgkQq7IHXV8xDPjdUgv/c4i/btU7rGFuJfqePeLjP09GDobJqlIWzHuB3wpBZa5s2oUqm9sTkbCZYW2KccD922rL6NrTVVq1vVC3y5YgDthCGGbm75GeYF95cJ8uDac4HEYDJVplgfZrvNXGuAUWJiHxtbU+1zBks5lculClNFw//3e+uLOzDJj52LjxZ88wSI+UCCI+yoY4hDyPCbJzVKb4zB7cclcuYNMTwMwP0fqo4dbdM8EQmejPUXESWS/s8MlWsbCjiym/kZSGvf8gskqWDLPMGrABWhbhGvKPdOZHhLI2IyS2TUBU2c93+t9WIV8xk76Ku5K1Tth4LXX07h63LWyIbwwLx5ayVqyzn0r9/4L/wjeM7Dz+OVnlM6kXs38Y3lBZncj5wuD1fVaZWYFIU77mfv/Vm9xVEIQ+anbtG2CYobsmunU/6swBvg6NZcjJCtINk3l/BE1xpwbOC6HSEw/qwRw0ldYySBpMijTUQL1OnD6ghQ0ZDgxGELQ9BQqhNNWR4ZDd4zxT4ius",
  "SHA256SUM": "e5872a199afeb919854938f89de402ff6f81237430e1ae22d2a7ad37432b1de5",
  "URL": "https://gitlab.manjaro.org/applications/calamares",
  "VERSION": "3.2.48-1",
}
`;
